### ISSUE TITLE
False Positive : unsupported-binary-operation when type hinting bool and None

### ISSUE THREAD

**Issue by Gornoka** [PyCQA/pylint#7478]
### Bug description

When executed on linux, pylint does not allow type hints with | operators.
Example file:

````python 
"""
pylint bug example
"""
import sys


def is_even(num: int) -> bool | None:
    """
    returns True if num is even, otherwise False,
    except when it is 0 then it returns None
    :param num:
    :type num:
    :return:
    :rtype:
    """
    if num == 0:
        return None

    return num % 2 == 0


def is_odd(num: int) -> int | None:
    """
    returns num if num is odd, otherwise None
    :param num:
    :type num:
    :return:
    :rtype:
    """
    if num == 0:
        return None

    return num % 2


if __name__ == '__main__':
    print(sys.version)
    print(is_even(2))
    print(is_even(3))
    print(is_odd(2))
    print(is_odd(3))

````

### Command used

```shell
python -m pylint .\pylint_bug_example.py
```

### Pylint output

```shell
output linux ( ubuntu in WSL) ( same for github action runners)

`
pylint pylint_bug_example.py 

************* Module pylint_bug_example
pylint_bug_example.py:7:25: E1131: unsupported operand type(s) for | (unsupported-binary-operation)
pylint_bug_example.py:22:24: E1131: unsupported operand type(s) for | (unsupported-binary-operation)

------------------------------------------------------------------
Your code has been rated at 2.31/10 (previous run: 1.54/10, +0.77)

`

Output Windows 
````shell
 python -m pylint .\pylint_bug_example.py

-------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 9.23/10, +0.77)
````
```


### Expected behavior

Windows and Linux pylint should behave equally, and the behavior should be as on windows, as the type hinting is indeed valid, and the program can be executed.

### Pylint version

```shell
### Linux
pylint 2.15.2
astroid 2.12.9
Python 3.8.10 (default, Jun 22 2022, 20:18:18)
[GCC 9.4.0]

### Windows

pylint 2.15.2
astroid 2.12.9
Python 3.10.0 (tags/v3.10.0:b494f59, Oct  4 2021, 19:00:18) [MSC v.1929 64 bit (AMD64)]
```

### OS / Environment

Windows11 python 3.10

Ubuntu 20.04 in wsl on windows and on github runner, if desired i can check bare metal next week

**Comment by Pierre-Sassoulas** [comment]
pylint is right here, you're analyzing your example with python 3.8 under linux, the ``|``operator for types union is not available until python 3.10 (unless you use ``from __future__ import annotations``). You could also use 3.10 for linux.

**Comment by Gornoka** [comment]
For anyone running into the same thing
The problem also existed on 3.10 systems because the RCfile was created at a python3.8 system back in the day,
 
The issue there was slightly different and not represented correctly by the description above.

The RCFILE has a `py-version` setting for the lowest python version to expect for version dependant features like typing, this appears to be created automatically when creating a new rcfile, with a fixed default value of the current python version at creation, instead of a blank value for a default of the current version at execution.

I would have expected  the `pylint --generate-rc` command to generate a blank / unset `py-version` value unless the current setting was set to a specific version by the user. 

@Pierre-Sassoulas do you think the above described pylint-rc creation behviour warrants an issue or is this intended behaviour ?

**Comment by DanielNoord** [comment]
This is "intended behaviour". `--generate-rcfile`, `--generate-toml-config` and `pylint-config` all use the current settings + default values. If you don't set `py-version` yourself then we take the default. This is the interpreter that is being used to run, like you expected. However, when we then "write down" this default value you get the issue where we use the interpreter that is being used to generate the configuration file.

I don't think we can solve for this as we can't detect when the value that is being used for `py-version` is coming from the user setting it specifically to the current interpreter or the current interpeter being used as default value.
We could add a comment about this to the description of the `py-version` option though.

**Comment by Pierre-Sassoulas** [comment]
I don't see a reasonable solution to this. We have an issue with the generated pylintrc not being the same depending on interpreter or enchant install so we can't check that it's up to date on the pipeline (can't link I'm on mobile), but I don't know if permitting to set the value to 'default' in the pylintrc is right. It would make the python version change depending on the interpreter which is the opposite of what I would want when a colleague tell me that 'she can't reproduce the error I told her about on her machine'. Debugging enchant install mismatch causing differing results also sound like hell on earth.

**Comment by Gornoka** [comment]
> I don't see a reasonable solution to this. We have an issue with the generated pylintrc not being the same depending on interpreter or enchant install so we can't check that it's up to date on the pipeline (can't link I'm on mobile), but I don't know if permitting to set the value to 'default' in the pylintrc is right. It would make the python version change depending on the interpreter which is the opposite of what I would want when a colleague tell me that 'she can't reproduce the error I told her about on her machine'. Debugging enchant install mismatch causing differing results also sound like hell on earth.

You are probably right, changing the entire config machinery to make one edge case more predictable is most definitely not worth the hassle.

> This is "intended behaviour". `--generate-rcfile`, `--generate-toml-config` and `pylint-config` all use the current settings + default values. If you don't set `py-version` yourself then we take the default. This is the interpreter that is being used to run, like you expected. However, when we then "write down" this default value you get the issue where we use the interpreter that is being used to generate the configuration file.
> 
> I don't think we can solve for this as we can't detect when the value that is being used for `py-version` is coming from the user setting it specifically to the current interpreter or the current interpeter being used as default value.
> We could add a comment about this to the description of the `py-version` option though.

I think a hint on the `--generate-rc` docs describing  which information will commonly be inferred from the environment would be nice, as this is not really obvious for new users. Are you aware of anything else that behaves like py-version?

**Comment by Pierre-Sassoulas** [comment]
Just a thought, maybe py-version's default should be 3.7.2 (the oldest interpreter pylint support).

The other example is the enchant dictionary used for spelling check.

**Comment by DanielNoord** [comment]
I think that might be annoying if you try to run `pylint` on a simple script/project without any setup or configuration files. I want to run that script on the interpreter I'm working on, not on the interpreter that one of my tools or IDE-integrated tools supports.

**Comment by Pierre-Sassoulas** [comment]
> annoying if you try to run pylint on a simple script/project without any setup or configuration files

Right, that would be very unfriendly to first time users.

I think we'll have to document better that when you generate some value will use your system defaults and you need to take special care before using it for your team.
